### PR TITLE
Decouple sequence-field's format from its in-memory types

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

Decouples the types used to represent sequence-field changes in memory from those we send over the wire. This change was relatively mechanical but is likely easier to review from a 'fresh code' standpoint, with the exception of moving the `DiscriminatedUnionDispatcher` into the `codec` library as it's a generally useful encoding for optional unions.

Notable features:
- Moves typebox schema types for sequence-field to a separate file from the in-memory types. This noticeably cleans up the dependency structure of sequence-field (only the codec file depends on the encoded format)
- More strongly-types the codec, avoiding using spread operators when encoding data.
- Leverages the discriminated union encoding outlined in `discriminatedUnions.ts` for marks

There are still some planned changes to make sequence-field's encoding more efficient, so this doesn't attempt to settle all of the encoding details. However, it should make future changes much more local, and also brings the organization of the code more inline with optional/value field.

## Breaking Changes

This is a breaking change for the persisted sequence-field format. No back-compat effort has been made.

### CODE REVIEW

File: experimental/dds/tree2/src/core/tree/persistedTreeTextFormat.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ interface EncodedGenericTreeNode<TChild>
 	extends EncodedGenericFieldsNode<TChild>,
 		EncodedNodeData {}
 const EncodedGenericTreeNode = <Schema extends TSchema>(tChild: Schema) =>
-	Type.Intersect([EncodedGenericFieldsNode(tChild), EncodedNodeData], {
+	Type.Composite([EncodedGenericFieldsNode(tChild), EncodedNodeData], {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/core/tree/persistedTreeTextFormat.ts:82]: I'm not sure why typebox's validator tolerated this before, but when using ajv this encoding complained on some optional field tests.

In general, `Type.Intersect` is a bad idea to use with `{ additionalProperties: false }`. The difference between the two is that `Type.Composite` constructs a single object type with the necessary fields, whereas `Type.Intersect` uses JSON schema's `allOf` specification.

**Comment by Abe27342** [comment]
the ajv dependency changes have become sprawling enough I'm going to pull those out into a separate PR.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts [hunk 2]
```diff
@@ -164,29 +61,14 @@ export const HasRevisionTag = Type.Object({ revision: Type.Optional(RevisionTagS
...
-	[
-		HasMoveId,
-		HasRevisionTag,
-		Type.Object({
-			type: Type.Literal("Insert"),
-		}),
-	],
-	noAdditionalProps,
-);
+const Insert = Type.Composite([HasMoveId, HasRevisionTag], noAdditionalProps);
```
Comment (Abe27342) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:64]: In general, not sure how I feel about duplicating comments in this file with the ones in `types`. We do expect the semantics of fields here to diverge with those in `types.ts` over time, but not sure what exactly is the most helpful to document.

I've used *some* discretion on which comments to preserve, but wasn't particularly careful about it.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts [hunk 3]
```diff
@@ -294,34 +143,69 @@ export interface MovePlaceholder extends HasRevisionTag, HasMoveId {
...
+		insert: Type.Optional(Insert),
+		moveIn: Type.Optional(MoveIn),
+		delete: Type.Optional(Delete),
+		moveOut: Type.Optional(MoveOut),
+		attachAndDetach: Type.Optional(AttachAndDetach),
+	},
+	unionOptions,
+);
+
+Type.Union([Attach, Detach, AttachAndDetach]);
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:163]: Remove?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldCodecs.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+			};
+			if (revision !== undefined) {
+				decoded.revision = revision;
+			}
+			if (adjacentCells !== undefined) {
+				// eslint-disable-next-line @typescript-eslint/no-shadow
+				decoded.adjacentCells = adjacentCells.map(([localId, count]) => ({
+					id: localId,
+					count,
+				}));
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldCodecs.ts:176]:
```suggestion
				decoded.adjacentCells = adjacentCells.map(([id, count]) => ({
					id,
					count,
				}));
```

**Review by yann-achard-MS — APPROVED** [review]

**Pull request merged**